File: day2/solution2.hs
import Data.List (sort)
import System.IO (IOMode (ReadMode), hGetContents, openFile)

-- Rock     1
-- Paper    2
-- Scisors  3

data Shape = None | Rock | Paper | Scisors deriving (Enum, Eq, Show)

data WinLoseTie = Win | Lose | Tie deriving (Eq, Show)

fromEnum' enum = case enum of
  Win -> 6
  Lose -> 0
  Tie -> 3

getShapes :: [String] -> [(Shape, WinLoseTie)]
getShapes = map getShapes
  where
    getShapes :: String -> (Shape, WinLoseTie)
    getShapes line = (getShape (head line), getWinLoseTie (last line))

    getShape :: Char -> Shape
    getShape 'A' = Rock
    getShape 'B' = Paper
    getShape 'C' = Scisors
    getShape _ = error "Not a valid Shape"

    getWinLoseTie :: Char -> WinLoseTie
    getWinLoseTie 'X' = Lose
    getWinLoseTie 'Y' = Tie
    getWinLoseTie 'Z' = Win
    getWinLoseTie _ = error "Not a valid WinLoseTie"

getScore :: (Shape, WinLoseTie) -> Int
getScore (a, b)
  | b == Tie = fromEnum' Tie + fromEnum a
  -- Rock first
  | a == Rock && b == Win = fromEnum' Win + fromEnum Paper
  | a == Rock && b == Lose = fromEnum Scisors
  -- Paper first
  | a == Paper && b == Win = fromEnum' Win + fromEnum Scisors
  | a == Paper && b == Lose = fromEnum Rock
  -- Scisors First
  | a == Scisors && b == Win = fromEnum' Win + fromEnum Rock
  | a == Scisors && b == Lose = fromEnum Paper
  | otherwise = error ("No case " ++ show (a, b))

main :: IO ()
main = do
  file <- openFile "input.txt" ReadMode
  content <- hGetContents file
  let fileLines = lines content

  putStr $ "The score is: " ++ show (sum . map getScore $ getShapes fileLines)

File: day2/solution2.py
with open("input.txt") as f:
    input = f.read()

import enum

# X LOSE
# Y DRAW
# Z WIN


class Shape(enum.Enum):
    ROCK = 1
    PAPER = 2
    SCISORS = 3


class WIN_DRAW(enum.Enum):
    LOSE = 0
    TIE = 3
    WIN = 6


shapes = []


for line in input.splitlines():
    a, b = line.split(" ")

    if a == "A":
        left = Shape.ROCK
    elif a == "B":
        left = Shape.PAPER
    elif a == "C":
        left = Shape.SCISORS
    else:
        raise RuntimeError("a: " + a)

    if b == "X":
        right = WIN_DRAW.LOSE
    elif b == "Y":
        right = WIN_DRAW.TIE
    elif b == "Z":
        right = WIN_DRAW.WIN
    else:
        raise RuntimeError("b: " + b)

    shapes.append((left, right))


score = 0

for left_shape, win_draw_tie in shapes:
    score += win_draw_tie.value

    if win_draw_tie == WIN_DRAW.TIE:
        shape = left_shape

    # find the shape
    elif left_shape == Shape.ROCK and win_draw_tie == WIN_DRAW.LOSE:
        shape = Shape.SCISORS
    elif left_shape == Shape.ROCK and win_draw_tie == WIN_DRAW.WIN:
        shape = Shape.PAPER

    elif left_shape == Shape.PAPER and win_draw_tie == WIN_DRAW.LOSE:
        shape = Shape.ROCK
    elif left_shape == Shape.PAPER and win_draw_tie == WIN_DRAW.WIN:
        shape = Shape.SCISORS

    elif left_shape == Shape.SCISORS and win_draw_tie == WIN_DRAW.LOSE:
        shape = Shape.PAPER
    elif left_shape == Shape.SCISORS and win_draw_tie == WIN_DRAW.WIN:
        shape = Shape.ROCK

    score += shape.value

print(score)
